### PR TITLE
fix(core): allow onDestroy unregistration while destroying

### LINKED ISSUE #50221: Erratic behavior with DestroyRef, not all callbacks are always being called upon component destruction.
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

We have migrated our code base to use the new takeUntilDestroyed pipe along with DestroyRef.

Unfortunately, we have noticed very erratic behavior that with these, going so far as to not actually unsubscribe from some Observables at all after components are destroyed (I am still working on trying to provide a minimal reproduction for some cases).

I have been able to replicate some bugs consistently with the reproduction below.
https://stackblitz.com/edit/at-ng-14-router-d9iur8?file=src%2Fapp%2Fhello.component.ts

# Reproduction
This small App contains
A Service with a BehaviorSubject (bhSubject) that increments an integer and a Subject (subject) that "increments" strings.
Those two actions are performed from AppComponent, which is always visible with the buttons.
A third button "Toggle Child" will show and hide the HelloComponent, which contains the actual bug.

HelloComponent registers a onDestroy callback in each of the following cases:
1) In the callback given to subscribe() on bhSubject.
2) In a method called from the same callback given to subscribe on bhSubject.
3) In the callback given to subscribe() on subject.
4) In a setTimeout of 0.
5) In ngZone.runOutsideAngular()
6) Directly in ngOnit()

Note that a lot of the stuff is going on in the console.

## Case 1:

- Toggle Child
- Toggle Child again

### Expected onDestroy callbacks to have been called:
- 1, 2, 4, 5?, 6

### Actual callbacks called
- 2, 4, 6

## Case 2:

- Toggle Child
- Int++
- Toggle Child

### Expected onDestroy callbacks to have been called:
- 1, 1, 2, 2, 4, 5?, 6

### Actual callbacks called
- 1, 2, 2, 4, 6

## Case 3:

- Toggle Child
- Letter++
- Toggle Child

### Expected onDestroy callbacks to have been called:
- 1, 2, 3, 4, 5?, 6

### Actual callbacks called
- 2, 3, 4, 6

Following this, it seems like the first callback from a BehaviorSubject doesn't allow to use DestroyRef.

Now you might be wondering why I added runOutsideAngular() in this, it's normal it's not working there. Well, if we comment the subscription to the normal subject (line 40 to 50 in hello.component.ts) and do case 1 again, it somehow does work with runOutsideAngular().

```ts
    /*this.someService.subject
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe((value) => {
        console.log(value);

        this.destroyRef.onDestroy(() => {
          console.log('Subject ' + value + ' destroyRef');
        });

        this.letterValue = value;
      });*/
```

## Case 4 (Same as Case 1, but with a subscription that is commented out):

- Toggle Child
- Toggle Child again

### Expected onDestroy callbacks to have been called:
- 1, 2, 4, 5?, 6

### Actual callbacks called
- 2, 4, 5, 6

This all seems very inconsistent. As stated before, we ran into bugs where takeUntilDestroyed() didn't work at all when used on observables within ngOnit directly (with destroyRef supplied, of course) and it wouldn't unsubscribe from all observables. Sadly I've not been able to reproduce this in a simple example yet, but this makes this new feature quite unusable. I'll keep you updated if I manage to craft a minimal reproduction for that bug as well.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/at-ng-14-router-d9iur8?file=src%2Fapp%2Fhello.component.ts

### Please provide the exception or error you saw

_No response_

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.0
Node: 16.18.0
Package Manager: npm 9.6.4
OS: win32 x64

Angular: 16.0.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, localize, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.0
@angular-devkit/build-angular   16.0.0
@angular-devkit/core            16.0.0
@angular-devkit/schematics      16.0.0
@schematics/angular             16.0.0
rxjs                            7.8.1
typescript                      5.0.4
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by garrettld**
Ensure that all onDestroy callbacks are called by allowing unregistering of onDestroy hooks while destroying.

Fixes #50221

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

`onDestroy` callbacks that unregister themselves when called will cause the next `onDestroy` callback to be skipped.

Issue Number: #50221

## What is the new behavior?
All `onDestroy` callbacks that are present when the destroy process starts will be called, regardless of any unregistration that may occur while destroying.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by garrettld** [comment]
@pkozlowski-opensource I went with setting the ON_DESTROY_HOOKS value to null (per your suggestion) to avoid the relatively more expensive approach of copying the array. Thanks for letting me take a crack at this.

File: packages/core/src/render3/node_manipulation.ts [hunk 1]
```diff
@@ -466,6 +466,7 @@ function processCleanups(tView: TView, lView: LView): void {
     lView[CLEANUP] = null;
   }
   const destroyHooks = lView[ON_DESTROY_HOOKS];
+  lView[ON_DESTROY_HOOKS] = null;
```
Comment (AndrewKushnir) [inline packages/core/src/render3/node_manipulation.ts:469]: It looks like there is a similar call below (at line 475), should we remove it from there? I think it'd also be helpful to add a comment into the code on why we reset it before invoking callbacks.

**Review by AndrewKushnir — COMMENTED** [review]
@garrettld thanks for the PR. I've just left a quick comment.

I'd also like to ask @pkozlowski-opensource to take a look, he is more familiar with this part of the code.

**Comment by garrettld** [comment]
@AndrewKushnir I removed the second `lView[ON_DESTROY_HOOKS] = null;`, I think it's relatively safe to do so since user code should not be able to register new hooks at this point since `LViewFlags.Destroyed` is already set. I also moved the reset into the `if (destroyHooks !== null)` block since it's only necessary if that condition is true.

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
> Thanks for letting me take a crack at this.

Thank YOU for contributing the fix, appreciate it!

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:531051235:BASE:531051261:1683765552595:8adbd681).

**Comment by pkozlowski-opensource** [comment]
@garrettld I've just realised that there is another code path that suffers from the same problem, this test (a bit convoluted, bit still pretty simple) exposes it:

```typescript
it('should allow unregistration while destroying', () => {
      const destroyedLog: string[] = [];

      const envInjector = createEnvironmentInjector([], TestBed.inject(EnvironmentInjector));
      const destroyRef = envInjector.get(DestroyRef);

      const unregister = destroyRef.onDestroy(() => {
        destroyedLog.push('first');
        unregister();
      });
      destroyRef.onDestroy(() => {
        destroyedLog.push('second');
      });

      expect(destroyedLog).toEqual([]);

      envInjector.destroy();
      expect(destroyedLog).toEqual(['first', 'second']);
    });
```

A similar fix needs to be done in https://github.com/angular/angular/blob/8ce99f5a0d958d933e4dda8036abc3683b61a9a7/packages/core/src/di/r3_injector.ts#L202 

Do you think you could update the PR to include the above test + fix for the other code path?

**Comment by pkozlowski-opensource** [comment]
Thnx for the update @garrettld - looks great!

**Comment by pkozlowski-opensource** [comment]
caretaker note: the latest [presubmit](https://fusion2.corp.google.com/presubmit/531187338/OCL:531187338:BASE:531187365:1683812853692:903635fb/targets) failures are pre-existing.

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 03f1e240b368c4c87fd8412605bf0eff9adc153d.

**Pull request closed without merging**